File: WinPaletter/Extras/ApplicationThemer.vb
Imports System.ComponentModel

Public Class ApplicationThemer

    Private BackupSettings As WPSettings
    Private _Shown As Boolean = False
    Private CloseAndApply As Boolean = False
    Public FixLanguageDarkModeBug As Boolean = True

    Private Sub ApplicationThemer_Editor_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        _Shown = False
        BackupSettings = New WPSettings(WPSettings.Mode.Registry)
        LoadLanguage
        ApplyStyle(Me)
        Button12.Image = MainFrm.Button20.Image.Resize(16, 16)
        ApplyFromCP(My.CP)
        AdjustPreview()
        CloseAndApply = False
    End Sub

    Private Sub ApplicationThemer_Shown(sender As Object, e As EventArgs) Handles Me.Shown
        _Shown = True
    End Sub

    Private Sub ApplicationThemer_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing

        If Not CloseAndApply Then 'Restore previous settings
            With My.Settings.Appearance
                .CustomColors = BackupSettings.Appearance.CustomColors
                .CustomTheme = BackupSettings.Appearance.CustomTheme
                .RoundedCorners = BackupSettings.Appearance.RoundedCorners
                .BackColor = BackupSettings.Appearance.BackColor
                .AccentColor = BackupSettings.Appearance.AccentColor
                .Save()
            End With

            FetchDarkMode()
            ApplyStyle()
        End If

        FixLanguageDarkModeBug = True
    End Sub

    Sub ApplyFromCP(CP As CP)
        With CP.AppTheme
            AppThemeEnabled.Checked = .Enabled
            appearance_dark.Checked = .DarkMode
            RoundedCorners.Checked = .RoundCorners
            BackColorPick.BackColor = .BackColor
            AccentColor.BackColor = .AccentColor
        End With

    End Sub

    Sub ApplyToCP(CP As CP)
        With CP.AppTheme
            .Enabled = AppThemeEnabled.Checked
            .DarkMode = appearance_dark.Checked
            .RoundCorners = RoundedCorners.Checked
            .BackColor = BackColorPick.BackColor
            .AccentColor = AccentColor.BackColor
        End With
    End Sub

    Sub AdjustPreview()
        If FixLanguageDarkModeBug Then Exit Sub

        With My.Settings.Appearance
            .CustomColors = True
            .CustomTheme = appearance_dark.Checked
            .RoundedCorners = RoundedCorners.Checked
            .BackColor = BackColorPick.BackColor
            .AccentColor = AccentColor.BackColor
        End With

        ApplyStyle(Me)

        For Each ctrl As Control In Controls
            ctrl.Invalidate()
        Next

        For Each ctrl As Control In GroupBox12.Controls
            ctrl.Invalidate()
        Next
    End Sub

    Private Sub Button11_Click(sender As Object, e As EventArgs) Handles Button11.Click
        If OpenFileDialog1.ShowDialog = DialogResult.OK Then
            Dim CPx As New CP(CP.CP_Type.File, OpenFileDialog1.FileName)
            ApplyFromCP(CPx)
            AdjustPreview()
            CPx.Dispose()
        End If
    End Sub

    Private Sub Button9_Click(sender As Object, e As EventArgs) Handles Button9.Click
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyFromCP(CPx)
        AdjustPreview()
        CPx.Dispose()
    End Sub

    Private Sub Button12_Click(sender As Object, e As EventArgs) Handles Button12.Click
        Using _Def As CP = CP_Defaults.From(My.PreviewStyle)
            ApplyFromCP(_Def)
            AdjustPreview()
        End Using
    End Sub

    Private Sub Button8_Click(sender As Object, e As EventArgs) Handles Button8.Click
        ApplyToCP(My.CP)
        CloseAndApply = False
        Close()
    End Sub

    Private Sub Button10_Click_1(sender As Object, e As EventArgs) Handles Button10.Click
        Cursor = Cursors.WaitCursor
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyToCP(CPx)
        ApplyToCP(My.CP)
        CPx.AppTheme.Apply()
        CloseAndApply = True
        BackupSettings = New WPSettings(WPSettings.Mode.Registry)
        CPx.Dispose()
        Cursor = Cursors.Default
    End Sub

    Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click
        CloseAndApply = False
        Close()
    End Sub

    Private Sub AppThemeEnabled_CheckedChanged(sender As Object, e As EventArgs) Handles AppThemeEnabled.CheckedChanged
        checker_img.Image = If(sender.Checked, My.Resources.checker_enabled, My.Resources.checker_disabled)
        AdjustPreview()
    End Sub

    Private Sub AccentColor_BackColorPick_DragDrop(sender As Object, e As DragEventArgs) Handles AccentColor.DragDrop, BackColorPick.DragDrop
        AdjustPreview()
    End Sub

    Private Sub AccentColor_Click(sender As Object, e As EventArgs) Handles AccentColor.Click

        If TypeOf e Is DragEventArgs Then Exit Sub

        With My.Settings.Appearance
            .CustomColors = BackupSettings.Appearance.CustomColors
            .CustomTheme = BackupSettings.Appearance.CustomTheme
            .RoundedCorners = BackupSettings.Appearance.RoundedCorners
            .BackColor = BackupSettings.Appearance.BackColor
            .AccentColor = BackupSettings.Appearance.AccentColor
        End With

        If DirectCast(e, MouseEventArgs).Button = MouseButtons.Right Then
            AccentColor.BackColor = SubMenu.ShowMenu(AccentColor)
            AdjustPreview()
            Exit Sub
        End If
        Dim clist As New List(Of Control) From {AccentColor}
        ColorPickerDlg.Pick(clist)
        clist.Clear()

        AdjustPreview()
    End Sub

    Private Sub BackColorPick_Click(sender As Object, e As EventArgs) Handles BackColorPick.Click
        With My.Settings.Appearance
            .CustomColors = BackupSettings.Appearance.CustomColors
            .CustomTheme = BackupSettings.Appearance.CustomTheme
            .RoundedCorners = BackupSettings.Appearance.RoundedCorners
            .BackColor = BackupSettings.Appearance.BackColor
            .AccentColor = BackupSettings.Appearance.AccentColor
        End With

        If DirectCast(e, MouseEventArgs).Button = MouseButtons.Right Then
            BackColorPick.BackColor = SubMenu.ShowMenu(BackColorPick)
            AdjustPreview()
            Exit Sub
        End If

        Dim clist As New List(Of Control) From {BackColorPick, Me}
        ColorPickerDlg.Pick(clist)
        clist.Clear()

        AdjustPreview()
    End Sub

    Private Sub CheckedChanged(sender As Object) Handles appearance_dark.CheckedChanged, RoundedCorners.CheckedChanged
        AdjustPreview()
    End Sub

    Private Sub Appearance_list_SelectedIndexChanged(sender As Object, e As EventArgs) Handles appearance_list.SelectedIndexChanged
        If _Shown Then
            Select Case appearance_list.SelectedItem.ToString.ToLower
                Case "Default Dark".ToLower
                    appearance_dark.Checked = True
                    RoundedCorners.Checked = (My.W11 Or My.W7)
                    AccentColor.BackColor = My.DefaultAccent
                    BackColorPick.BackColor = My.DefaultBackColorDark

                Case "Default Light".ToLower
                    appearance_dark.Checked = False
                    RoundedCorners.Checked = (My.W11 Or My.W7)
                    AccentColor.BackColor = My.DefaultAccent
                    BackColorPick.BackColor = My.DefaultBackColorLight

                Case "AMOLED".ToLower
                    appearance_dark.Checked = True
                    RoundedCorners.Checked = (My.W11 Or My.W7)
                    AccentColor.BackColor = My.DefaultAccent
                    BackColorPick.BackColor = Color.Black

                Case "Extreme White".ToLower
                    appearance_dark.Checked = False
                    RoundedCorners.Checked = (My.W11 Or My.W7)
                    AccentColor.BackColor = My.DefaultAccent
                    BackColorPick.BackColor = Color.White

                Case "GitHub Dark".ToLower
                    appearance_dark.Checked = True
                    RoundedCorners.Checked = True
                    AccentColor.BackColor = Color.FromArgb(19, 35, 58)
                    BackColorPick.BackColor = Color.FromArgb(13, 17, 23)

                Case "GitHub Light".ToLower
                    appearance_dark.Checked = False
                    RoundedCorners.Checked = True
                    AccentColor.BackColor = Color.FromArgb(31, 111, 235)
                    BackColorPick.BackColor = Color.FromArgb(246, 248, 250)

                Case "Reddit Dark".ToLower
                    appearance_dark.Checked = True
                    RoundedCorners.Checked = True
                    AccentColor.BackColor = Color.FromArgb(255, 70, 0)
                    BackColorPick.BackColor = Color.FromArgb(9, 9, 9)

                Case "Reddit Light".ToLower
                    appearance_dark.Checked = False
                    RoundedCorners.Checked = True
                    AccentColor.BackColor = Color.FromArgb(255, 70, 0)
                    BackColorPick.BackColor = Color.FromArgb(242, 242, 242)

                Case "Discord Dark".ToLower
                    appearance_dark.Checked = True
                    RoundedCorners.Checked = False
                    AccentColor.BackColor = Color.FromArgb(65, 71, 78)
                    BackColorPick.BackColor = Color.FromArgb(32, 34, 38)

                Case "Discord Light".ToLower
                    appearance_dark.Checked = False
                    RoundedCorners.Checked = False
                    AccentColor.BackColor = Color.FromArgb(138, 140, 143)
                    BackColorPick.BackColor = Color.FromArgb(255, 255, 255)

            End Select

            AdjustPreview()
        End If
    End Sub

    Private Sub ApplicationThemer_BackColorChanged(sender As Object, e As EventArgs) Handles Me.BackColorChanged
        For Each ctrl As Control In Controls
            ctrl.Invalidate()
        Next

        For Each ctrl As Control In GroupBox12.Controls
            ctrl.Invalidate()
        Next
    End Sub

    Private Sub Form_HelpButtonClicked(sender As Object, e As CancelEventArgs) Handles Me.HelpButtonClicked
        Process.Start(My.Resources.Link_Wiki & "/Edit-WinPaletter-application-theme")
    End Sub

End Class

File: WinPaletter/Extras/ApplicationThemer.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;

namespace WinPaletter
{
    public partial class ApplicationThemer
    {
        public bool FixLanguageDarkModeBug = true;

        private void Form_HelpButtonClicked(object sender, CancelEventArgs e)
        {
            Process.Start($"{Properties.Resources.Link_Wiki}/Edit-WinPaletter-application-theme");
        }

        public ApplicationThemer()
        {
            InitializeComponent();
        }

        private void LoadFromWPTH(object sender, EventArgs e)
        {
            using (OpenFileDialog dlg = new() { Filter = Program.Filters.WinPaletterTheme, Title = Program.Lang.Filter_OpenWinPaletterTheme })
            {
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    using (Theme.Manager TMx = new(Theme.Manager.Source.File, dlg.FileName))
                    {
                        LoadFromTM(TMx);
                    }
                }
            }
        }

        private void LoadFromCurrent(object sender, EventArgs e)
        {
            Theme.Manager TMx = new(Theme.Manager.Source.Registry);
            LoadFromTM(TMx);
            TMx.Dispose();
        }

        private void LoadFromDefault(object sender, EventArgs e)
        {
            Theme.Manager TMx = Theme.Default.Get(Program.WindowStyle);
            LoadFromTM(TMx);
            TMx.Dispose();
        }

        private void LoadIntoCurrentTheme(object sender, EventArgs e)
        {
            ApplyToTM(Program.TM);
            Close();
        }
        private void Apply(object sender, EventArgs e)
        {
            Cursor = Cursors.WaitCursor;

            using (Theme.Manager TMx = new(Theme.Manager.Source.Registry))
            {
                ApplyToTM(TMx);
                ApplyToTM(Program.TM);
                TMx.AppTheme.Apply();
            }

            Cursor = Cursors.Default;
        }

        private void ApplicationThemer_Editor_Load(object sender, EventArgs e)
        {
            DesignerData data = new(this)
            {
                AspectName = Program.Lang.Store_Toggle_AppTheme,
                Enabled = Program.TM.AppTheme.Enabled,
                Import_theme = false,
                Import_msstyles = false,
                GeneratePalette = false,
                GenerateMSTheme = false,
                Import_preset = false,
                CanSwitchMode = false,

                OnLoadIntoCurrentTheme = LoadIntoCurrentTheme,
                OnApply = Apply,
                OnImportFromDefault = LoadFromDefault,
                OnImportFromWPTH = LoadFromWPTH,
                OnImportFromCurrentApplied = LoadFromCurrent,
            };

            LoadData(data);

            LoadFromTM(Program.TM);
            AdjustPreview();
        }

        private void ApplicationThemer_FormClosing(object sender, FormClosingEventArgs e)
        {
            FixLanguageDarkModeBug = true;
        }

        public void LoadFromTM(Theme.Manager TM)
        {
            AspectEnabled = TM.AppTheme.Enabled;
            appearance_dark.Checked = TM.AppTheme.DarkMode;
            RoundedCorners.Checked = TM.AppTheme.RoundCorners;
            BackColorPick.BackColor = TM.AppTheme.BackColor;
            AccentColor.BackColor = TM.AppTheme.AccentColor;
            checkBox1.Checked = TM.AppTheme.Animations;
            colorItem1.BackColor = TM.AppTheme.SecondaryColor;
            colorItem2.BackColor = TM.AppTheme.TertiaryColor;
            colorItem3.BackColor = TM.AppTheme.DisabledColor;
            colorItem4.BackColor = TM.AppTheme.DisabledBackColor;
        }

        public void ApplyToTM(Theme.Manager TM)
        {
            TM.AppTheme.Enabled = AspectEnabled;
            TM.AppTheme.DarkMode = appearance_dark.Checked;
            TM.AppTheme.RoundCorners = RoundedCorners.Checked;
            TM.AppTheme.BackColor = BackColorPick.BackColor;
            TM.AppTheme.AccentColor = AccentColor.BackColor;
            TM.AppTheme.Animations = checkBox1.Checked;
            TM.AppTheme.SecondaryColor = colorItem1.BackColor;
            TM.AppTheme.TertiaryColor = colorItem2.BackColor;
            TM.AppTheme.DisabledColor = colorItem3.BackColor;
            TM.AppTheme.DisabledBackColor = colorItem4.BackColor;
        }

        public void AdjustPreview()
        {
            if (FixLanguageDarkModeBug)
                return;

            Config.Scheme scheme = new(AccentColor.BackColor, BackColorPick.BackColor, appearance_dark.Checked);
            testControl1.Scheme = scheme;
            testControl4.Scheme = scheme;
            testControl5.Scheme = scheme;
            testControl2.Scheme = scheme;
            testControl3.Scheme = scheme;

            Config.Scheme scheme_sec = new(colorItem1.BackColor, BackColorPick.BackColor, appearance_dark.Checked);
            testControl7.Scheme = scheme_sec;
            testControl6.Scheme = scheme_sec;

            Config.Scheme scheme_ter = new(colorItem2.BackColor, BackColorPick.BackColor, appearance_dark.Checked);
            testControl9.Scheme = scheme_ter;
            testControl8.Scheme = scheme_ter;

            Config.Scheme scheme_dis = new(colorItem3.BackColor, colorItem4.BackColor, appearance_dark.Checked);
            testControl14.Scheme = scheme_dis;
            testControl11.Scheme = scheme_dis;
            testControl10.Scheme = scheme_dis;
            testControl13.Scheme = scheme_dis;
            testControl12.Scheme = scheme_dis;


            foreach (Control ctrl in Controls)
                ctrl.Invalidate();

            foreach (Control ctrl in titlebarExtender1.Controls)
                ctrl.Invalidate();
        }

        private void AccentColor_BackColorPick_DragDrop(object sender, DragEventArgs e)
        {
            AdjustPreview();
        }

        private void CheckedChanged(object sender, EventArgs e)
        {
            AdjustPreview();
        }

        private void Appearance_list_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (IsShown)
            {
                switch (appearance_list.SelectedItem.ToString().ToLower() ?? string.Empty)
                {
                    case var @case when @case == ("Default Dark".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = true;
                            RoundedCorners.Checked = OS.W12 || OS.W11 || OS.W7;
                            AccentColor.BackColor = DefaultColors.PrimaryColor;
                            BackColorPick.BackColor = DefaultColors.BackColorDark;
                            break;
                        }

                    case var case1 when case1 == ("Default Light".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = false;
                            RoundedCorners.Checked = OS.W12 || OS.W11 || OS.W7;
                            AccentColor.BackColor = DefaultColors.PrimaryColor;
                            BackColorPick.BackColor = DefaultColors.BackColorLight;
                            break;
                        }

                    case var case2 when case2 == ("AMOLED".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = true;
                            RoundedCorners.Checked = OS.W12 || OS.W11 || OS.W7;
                            AccentColor.BackColor = Color.FromArgb(0, 77, 193);
                            BackColorPick.BackColor = Color.Black;
                            break;
                        }

                    case var case3 when case3 == ("Extreme White".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = false;
                            RoundedCorners.Checked = OS.W12 || OS.W11 || OS.W7;
                            AccentColor.BackColor = DefaultColors.PrimaryColor;
                            BackColorPick.BackColor = Color.White;
                            break;
                        }

                    case var case4 when case4 == ("GitHub Dark".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = true;
                            RoundedCorners.Checked = true;
                            AccentColor.BackColor = Color.FromArgb(19, 35, 58);
                            BackColorPick.BackColor = Color.FromArgb(13, 17, 23);
                            break;
                        }

                    case var case5 when case5 == ("GitHub Light".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = false;
                            RoundedCorners.Checked = true;
                            AccentColor.BackColor = Color.FromArgb(31, 111, 235);
                            BackColorPick.BackColor = Color.FromArgb(246, 248, 250);
                            break;
                        }

                    case var case6 when case6 == ("Reddit Dark".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = true;
                            RoundedCorners.Checked = true;
                            AccentColor.BackColor = Color.FromArgb(255, 70, 0);
                            BackColorPick.BackColor = Color.FromArgb(9, 9, 9);
                            break;
                        }

                    case var case7 when case7 == ("Reddit Light".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = false;
                            RoundedCorners.Checked = true;
                            AccentColor.BackColor = Color.FromArgb(216, 49, 3);
                            BackColorPick.BackColor = Color.FromArgb(242, 242, 242);
                            break;
                        }

                    case var case8 when case8 == ("Discord Dark".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = true;
                            RoundedCorners.Checked = false;
                            AccentColor.BackColor = Color.FromArgb(35, 69, 117);
                            BackColorPick.BackColor = Color.FromArgb(32, 34, 38);
                            break;
                        }

                    case var case9 when case9 == ("Discord Light".ToLower() ?? string.Empty):
                        {
                            appearance_dark.Checked = false;
                            RoundedCorners.Checked = false;
                            AccentColor.BackColor = Color.FromArgb(130, 132, 135);
                            BackColorPick.BackColor = Color.FromArgb(255, 255, 255);
                            break;
                        }

                }

                AdjustPreview();
            }
        }

        private void AccentColor_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
            {
                AdjustPreview();
                return;
            }

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                AccentColor.BackColor = Forms.SubMenu.ShowMenu(AccentColor);
                AdjustPreview();
                return;
            }

            Dictionary<Control, string[]> CList = new()
            {
                { AccentColor, new string[] { nameof(AccentColor.BackColor) } }
            };

            Forms.ColorPickerDlg.Pick(CList);
            CList.Clear();

            AdjustPreview();
        }

        private void BackColorPick_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
            {
                AdjustPreview();
                return;
            }

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                BackColorPick.BackColor = Forms.SubMenu.ShowMenu(BackColorPick);
                AdjustPreview();
                return;
            }

            Dictionary<Control, string[]> CList = new()
            {
                { BackColorPick, new string[] { nameof(BackColorPick.BackColor) } }
            };

            Forms.ColorPickerDlg.Pick(CList);
            CList.Clear();

            AdjustPreview();
        }

        private void colorItem1_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
            {
                AdjustPreview();
                return;
            }

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                colorItem1.BackColor = Forms.SubMenu.ShowMenu(colorItem1);
                AdjustPreview();
                return;
            }

            Dictionary<Control, string[]> CList = new()
            {
                { colorItem1, new string[] { nameof(colorItem1.BackColor) } }
            };

            Forms.ColorPickerDlg.Pick(CList);
            CList.Clear();

            AdjustPreview();
        }

        private void colorItem2_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
            {
                AdjustPreview();
                return;
            }

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                colorItem2.BackColor = Forms.SubMenu.ShowMenu(colorItem2);
                AdjustPreview();
                return;
            }

            Dictionary<Control, string[]> CList = new()
            {
                { colorItem2, new string[] { nameof(colorItem2.BackColor) } }
            };

            Forms.ColorPickerDlg.Pick(CList);
            CList.Clear();

            AdjustPreview();
        }

        private void colorItem3_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
            {
                AdjustPreview();
                return;
            }

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                colorItem3.BackColor = Forms.SubMenu.ShowMenu(colorItem3);
                AdjustPreview();
                return;
            }

            Dictionary<Control, string[]> CList = new()
            {
                { colorItem3, new string[] { nameof(colorItem3.BackColor) } }
            };

            Forms.ColorPickerDlg.Pick(CList);
            CList.Clear();

            AdjustPreview();
        }

        private void colorItem4_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
            {
                AdjustPreview();
                return;
            }

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                colorItem4.BackColor = Forms.SubMenu.ShowMenu(colorItem4);
                AdjustPreview();
                return;
            }

            Dictionary<Control, string[]> CList = new()
            {
                { colorItem4, new string[] { nameof(colorItem4.BackColor) } }
            };

            Forms.ColorPickerDlg.Pick(CList);
            CList.Clear();

            AdjustPreview();
        }
    }
}